src/cls/base.ts

```diff
@@ -0,0 +1,95 @@
+/**
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {EventEmitter} from 'events';
+
+// tslint:disable-next-line:no-any
+export type Func<T = void> = (...args: any[]) => T;
+
+/**
+ * An interface that represents a background mechanism which is capable of
+ * storing, propagating and retrieving arbitrary continuation-local data (also
+ * called "context").
+ *
+ * A continuation refers to a logical tree of execution paths; a function passed
+ * to CLS#runWithCurrentContext is considered the root of such a tree, and
+ * subsequent child nodes are functions that are "triggered" because of
+ * asynchronous operations started by their parent (for example, a function that
+ * calls fs.readFile "triggers" the callback that is passed to it).
+ * The exact definition of "trigger" is implementation-dependent.
+ *
+ * CLS stands for continuation-local storage.
+ */
+export interface CLS<Context extends {}> {
```

Comment:
Based on the API, I don't think this is an interface for continuation local _storage_. I would have expected that to have a general `get`/`set` interface for key value pairs. This interface only offers a context. Would be more appropriate to call this `ContinuationContext`?

Secondly, please document the `Context` template parameter. I think it would be more appropriate to call it `ContextType`.

Reply:
It's continuation-local storage under an implicit global key. I understand the confusion though. I can do a refactor at the end, if you have no further comments.